### ISSUE TITLE
QuickestPath stops with AP Dump

### ISSUE THREAD

**Issue by chraibi** [JuPedSim/jpscore#195]
In Gitlab by @chraibi on Mar 3, 2016, 11:35 [[origin](https://gitlab.version.fz-juelich.de/jupedsim/jpscore/issues/181)]

Quickest Path stops: We think, that it happens in special cases only. One of these special cases is described now:
A subroom consists of a closed door and an open door.

Pedestrians, that start in that subroom, manage to get out.

Pedestrians, that get pushed in that subroom (via the closed door), will not find a way out.

Such a room is Room 9, Subroom 1

**Reproduction Steps:**

1. run with provided ini, geo and routing file

**Expected behavior:**

Even if a pedestrian is pushed into another (sub-)room, the router should find an exit

**Observed behavior:**

Dump AP gets called with msg: 
~~~
ERROR:   Cannot find valid destination for ped [1174] located in room [9] subroom [1] going to destination [-1]
ERROR:  No route to destination  [ -1 ]
ERROR:  Check your configuration file
~~~

**Screenshots and GIFs**
~~~
Time:  54.30 s | Evacuated:    91 / 1660 (  5% ) [==>                                      ]
ERROR:   Cannot find valid destination for ped [1174] located in room [9] subroom [1] going to destination [-1]
ERROR:  No route to destination  [ -1 ]
ERROR:  Check your configuration file
--------> Dumping AP <-----------

 Friendly ID: trans_26_room_9_subroom_1 centre = [ 16.6103, 81.6841 ]
 Real ID: 1286
 Length:  1.00012
 Is final exit to outside :0
 Distance to final goals


 transit to final goals:

         to UID ---> [ -1 ]      ---> [ Nothing ]

 connected to aps : 

 queue [  ]


------------------------------
~~~

![Screenshots and GIFs which follow reproduction steps to demonstrate the problem](url)

- **JuPedSim version:** v0.8
- **OS and version:** Mac/Linux
- **Compiler and version:** gcc 4.8 / clang 3.1

**Additional Info**

* Problem happens with the attached ini and geometry files: [Yes]
* Might be connected to my latest change of subroom ID to UID in some parts (plz be very careful)
* GetBestDefaultRandomExit

[aaaaQuickest_UBF_id_1_seed_1905.xml.zip](https://gitlab.version.fz-juelich.de/JuPedSim/jpscore/uploads/d782a4683257707a9cae839b38567fc0/aaaaQuickest_UBF_id_1_seed_1905.xml.zip)

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

closed

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by chraibi on 03 Mar 2016, 11:24_

---

+1

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by Arne on 03 Mar 2016, 12:11_

---

while on this bug, we can look into: GlobalRouter:984
~~~
     //fallback
     if(relevantAPS.size()==0)
     {
          //fixme: this should also never happened. But hapen due to previvous bugs..
          const vector<int>& goals=sub->GetAllGoalIDs();
          for(unsigned int g1=0; g1<goals.size(); g1++)
          {
               relevantAPS.push_back(_accessPoints[goals[g1]]);
          }

     }
~~~

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by OSchmidts on 07 Mar 2016, 08:42_

---

Seems to be a multithreading problem.

Reason for occurence:
after a pedestrian is pushed out of his (sub)room he should be corrected in Simulation::UpdateRoutesAndLocations however he is not due to the fact that GlobalRouter::GetRelevantRoutesTofinalDestination goes into the fallback routine and fails because the pedestrian is not in the subroom he thinks he is

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by Arne on 07 Mar 2016, 15:01_

---

Function **NormalSubroom::IsInSubRoom(...)** is not 3d proof!!

We cannot loop thru all subrooms and check **IsInSubRoom**, cause it will give false positives.

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by Arne on 07 Mar 2016, 15:06_

---

Speed check:
~~~
double normVi = ped->GetV().ScalarProduct(ped->GetV()); //squared
double HighVel = (ped->GetV0Norm() + delta) * (ped->GetV0Norm() + delta); //(v0+delta)^2
if (normVi > HighVel && ped->GetV0Norm() > 0)
~~~
Will this work, even if a ped is changing from a room with no elevation to a room **with** elevation?

I think, the ``ped->GetV()`` / ``ped->GetV0Norm()`` will crash the check??

(at least if either one gives speed of previous time step)

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by OSchmidts on 07 Mar 2016, 15:15_

---

@schroederbenjamin i tried to fix this, pls check (branch issue 184).
It is always working with one core so if a student need this desperatly and my fix is not working try as single core

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by b.schroeder on 07 Mar 2016, 16:34_

---

@OSchmidts I guess you mean Issue 181 right? I'll check that

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by OSchmidts on 08 Mar 2016, 07:39_

---

you are right. I am excited to see your results.

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by b.schroeder on 08 Mar 2016, 12:59_

---

Great work @OSchmidts it seems to work. Even on JURECA in parallel.

I'll make some more calculations in order to check the robustness

![Quickest_UBF_id_1_traj_seed_1905_travisto_snap_160308_13h_0](https://gitlab.version.fz-juelich.de/JuPedSim/jpscore/uploads/c2c657ebbc628f747b7271ed92123ac4/Quickest_UBF_id_1_traj_seed_1905_travisto_snap_160308_13h_0.png)

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by OSchmidts on 08 Mar 2016, 13:13_

---

nice, you can also try to remove 
~~~.cpp
#pragma omp barrier
~~~
in `Simulation::UpdateRoutesAndLocations() line 361` and see if it still works. This would do me a favour because i am back to research center only on monday.

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by OSchmidts on 21 Mar 2016, 13:07_

---

@schroederbenjamin @chraibi @Arne 
i think i fixed it (now for real, tested with a lot of different seeds)
please validate the fix! (branch issue_181)

edit: nevermind :(

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by chraibi on 18 Jul 2016, 11:16_

---

This bug is still _alive_.

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by chraibi on 18 Jul 2016, 15:44_

---

I just run this simulation 10 times on hermes. The error can not be reproduced. 

However, there are tones of the following WARNINGS and ERRORS
![Screen_Shot_2016-07-18_at_17.42.40](https://gitlab.version.fz-juelich.de/JuPedSim/jpscore/uploads/fce57c6d45bb1272c2371166c17ca7a6/Screen_Shot_2016-07-18_at_17.42.40.png)

that should be inspected and fixed, but in another issue.

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Dec 17, 2016, 12:08

_Original comment by jenkins on 18 Jul 2016, 15:49_

---

mentioned in commit f89c76b9534254c4043fa78c9ff64e80de36476b

**Comment by chraibi** [comment]
In Gitlab by @chraibi on Jun 13, 2018, 16:41

removed milestone
